pkg/util/winutil/pdhutil/pdhformatter.go

```diff
@@ -0,0 +1,130 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+// +build windows
+
+package pdhutil
+
+import (
+	"fmt"
+	"reflect"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+// PdhFormatter implements a formatter for PDH performance counters
+type PdhFormatter struct {
+	buf []uint8
+}
+
+// PdhCounterValue represents a counter value
+type PdhCounterValue struct {
+	Format  uint32
+	CStatus uint32
+	Double  float64
+	Large   int64
+	Long    int32
+}
+
+// ValueEnumFunc implements a callback for counter enumeration
+type ValueEnumFunc func(s string, v PdhCounterValue)
+
+// Enum enumerates performance counter values for a wildcard instance counter (e.g. `\Process(*)\% Processor Time`)
+func (f *PdhFormatter) Enum(hCounter PDH_HCOUNTER, format uint32, fn ValueEnumFunc) error {
+	var bufLen uint32
+	var itemCount uint32
+	r, _, _ := procPdhGetFormattedCounterArray.Call(
+		uintptr(hCounter),
+		uintptr(format),
+		uintptr(unsafe.Pointer(&bufLen)),
+		uintptr(unsafe.Pointer(&itemCount)),
+		uintptr(0),
+	)
+
+	if r != PDH_MORE_DATA {
+		return fmt.Errorf("Failed to get formatted counter array buffer size 0x%x", r)
+	}
+
+	if bufLen > uint32(len(f.buf)) {
+		f.buf = make([]uint8, bufLen)
+	}
+
+	buf := f.buf[:bufLen]
+
+	r, _, _ = procPdhGetFormattedCounterArray.Call(
+		uintptr(hCounter),
+		uintptr(format),
+		uintptr(unsafe.Pointer(&bufLen)),
+		uintptr(unsafe.Pointer(&itemCount)),
+		uintptr(unsafe.Pointer(&buf[0])),
+	)
+	if r != ERROR_SUCCESS {
+		return fmt.Errorf("Error getting formatted counter array 0x%x", r)
+	}
+
+	var items []PDH_FMT_COUNTERVALUE_ITEM
+	hdrItems := (*reflect.SliceHeader)(unsafe.Pointer(&items))
```

Comment:
The inner rust developer in me sets off alarms whenever I see an unsafe pointer dereference. Is it possible for any part of this to fail?

I would add a comment explaining why this is safe to do, and how we know that we won't segfault whenever you use unsafe.

Reply:
Adding a comment sounds like a good idea.